Make hassfest.dependencies faster with multiprocessing

akx opened the pull request:
## Proposed change

This PR makes `hassfest.dependencies` faster by using `multiprocessing` for the slow part (which is parsing all integration source code).

The speedup is about 60% on my macOS machine, but will depend on the number of cores available, of course:

```
(dev) $ hyperfine 'python -m script.hassfest -p dependencies'
Benchmark 1: python -m script.hassfest -p dependencies
  Time (mean ± σ):     10.798 s ±  1.204 s    [User: 9.569 s, System: 0.582 s]
  Range (min … max):    9.564 s … 13.960 s    10 runs
(faster-hassfest-deps) $ hyperfine 'python -m script.hassfest -p dependencies'
Benchmark 1: python -m script.hassfest -p dependencies
  Time (mean ± σ):      4.449 s ±  0.107 s    [User: 38.233 s, System: 4.630 s]
  Range (min … max):    4.299 s …  4.645 s    10 runs
```

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
  - No new tests, `hassfest` is validated by... everything really.

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

akx committed:
hassfest.dependencies: split to two loops

akx committed:
hassfest.dependencies: use multiprocessing for import scan

balloob commented:
Current version of this check runs in 4.5s on my M1 macbook.

Instead of wrapping it all in a pool, it would be better to find what is so slow on your end. Would it be the check if a file/dir exists:

https://github.com/akx/home-assistant-core/blob/faster-hassfest-deps/script/hassfest/dependencies.py#L211-L213

And if so, could that be done in a single fetch of the whole directly listing instead of doing individual checks.

akx commented:
@balloob Lucky you – I'm still running a 2019 clunker with "only" a 2,4 GHz 8-Core Core i9... 😁 

I did run a profiler, of course, before embarking on optimizing things. By far the most time spent (in a profiler run, so the times shown in the screenshot are inflated), is the `Collector.collect` hierarchy that does `parse` for each file and a whole lot of AST visitation.

This screenshot is from the PyCharm profiler's output at 9c85d22bab333556efd4194fc74842b044412cfe.

![home-assistant-core14](https://user-images.githubusercontent.com/58669/199785814-762eb2c0-5fc2-4f2e-bdd5-4f5350cf0122.png)

(There's no point comparing the same diagram post-multiprocessing-enablement, since the child processes aren't slowed down by the profiler.)

akx commented:
@balloob Ah, actually the https://github.com/akx/home-assistant-core/blob/faster-hassfest-deps/script/hassfest/dependencies.py#L211-L213 branch is never even taken even on `dev` (see code coverage gutter being red), so that won't be a source of slowness :)

<img width="704" alt="Screenshot 2022-11-03 at 19 16 28" src="https://user-images.githubusercontent.com/58669/199790219-a5a5b1ec-ffbd-4fad-be60-7d9a0bdc2191.png">

Rudd-O reviewed (approved):
I like this (it would help me and many others) and it appears correct.

akx commented on script/hassfest/dependencies.py:
```diff
@@ -218,42 +219,54 @@ def find_non_referenced_integrations(
+
+    # Find integration dependencies with multiprocessing
+    # (because it takes some time to parse thousands of files)
+    with multiprocessing.Pool() as pool:
```
I tried, and no, it's not. The time is spent mostly in Python visitor code, which is still GIL-locked.

akx commented:
@balloob Have you had a chance to look at the discussion above? I rebased this branch, and the perf increase seems relevant still.

In GitHub Actions, the `Check hassfest` action says
* [`dev`](https://github.com/home-assistant/core/actions/runs/3479878171/jobs/5818987578): `Validating dependencies... done in 12.41s`
* this PR: `Validating dependencies... done in 5.41s`

Locally: 

```
$ git checkout dev && hyperfine 'python -m script.hassfest -p dependencies' && git checkout faster-hassfest-deps && Switched to branch 'dev'
Your branch is up to date with 'origin/dev'.
hyperfine 'python -m script.hassfest -p dependencies'
Benchmark 1: python -m script.hassfest -p dependencies
  Time (mean ± σ):     14.578 s ±  0.846 s    [User: 13.019 s, System: 0.951 s]
  Range (min … max):   13.863 s … 16.779 s    10 runs

Switched to branch 'faster-hassfest-deps'
Your branch is up to date with 'origin/faster-hassfest-deps'.
Benchmark 1: python -m script.hassfest -p dependencies
  Time (mean ± σ):      4.612 s ±  0.172 s    [User: 36.067 s, System: 5.783 s]
  Range (min … max):    4.418 s …  4.925 s    10 runs
$
```

akx commented:
Rebased once again, on top of a1a324a02eeee25bfc81309004bbbefa99aab0c3 now 😄 

This is still ~3x faster than `dev` on my box.

```
$ git checkout dev && hyperfine 'python -m script.hassfest -p dependencies' && git checkout faster-hassfest-deps && hyperfine 'python -m script.hassfest -p dependencies'
Switched to branch 'dev'
Your branch is up to date with 'upstream/dev'.
Benchmark 1: python -m script.hassfest -p dependencies
  Time (mean ± σ):      8.683 s ±  0.424 s    [User: 8.138 s, System: 0.477 s]
  Range (min … max):    8.395 s …  9.535 s    10 runs

Switched to branch 'faster-hassfest-deps'
Your branch is up to date with 'origin/faster-hassfest-deps'.
Benchmark 1: python -m script.hassfest -p dependencies
  Time (mean ± σ):      2.969 s ±  0.153 s    [User: 25.163 s, System: 2.940 s]
  Range (min … max):    2.766 s …  3.183 s    10 runs
```

akx commented:
Could e.g. @frenck take a look at this? 😄

emontnemery commented on script/hassfest/dependencies.py:
```diff
@@ -297,17 +313,21 @@ def _check_circular_deps(
 
-        if config.specific_integrations:
+
+def _validate_dependencies_exist(
```
The function checks both that referenced dependencies exist and that they're not duplicated in dependencies and after_dependencies. Please update the function name and the comment accordingly

emontnemery reviewed (approved):
This is a great speedup 👍
Just a minor remark about function naming, which does not block merging the PR

akx commented:
A mere 146 days later! 😉 😁 

Thanks for the merge.
